Backport fixes for ARM64EC import libraries

dpaoliello opened the pull request:
Backports fixes to be able to generate import libraries for ARM64EC:
* <https://github.com/llvm/llvm-project/pull/78769>
* <https://github.com/llvm/llvm-project/pull/81059>
* <https://github.com/llvm/llvm-project/pull/78772>
* <https://github.com/llvm/llvm-project/pull/84834>

There is one modification to these commits: we removed support for `EXPORTAS` in def files to maintain ABI compatibility for the `COFFShortExport` struct.

cjacek committed:
[llvm-readobj][Object][COFF] Print COFF import library symbol export name. (#78769)

getExportName implementation is based on lld-link. In its current form,
it's mostly about convenience, but it will be more useful for EXPORTAS
support, for which export name is not possible to deduce from other
printed properties.

cjacek committed:
[llvm-lib][llvm-dlltool][Object] Add support for EXPORTAS name types. (#78772)

EXPORTAS is a new name type in import libraries. It's used by default on ARM64EC,
but it's allowed on other platforms as well.

cjacek committed:
[llvm-lib][Object] Add support for EC importlib symbols. (#81059)

ARM64EC import libraries expose two additional symbols: mangled thunk
symbol (like `#func`) and auxiliary import symbol (like`__imp_aux_func`).
The main functional change with this patch is that those symbols are
properly added to static library ECSYMBOLS.

efriedma-quic commented:
This seems like a pretty big change to backport... how useful is it in practice?  I was under the impression that arm64ec lld support is still immature... and if you're using the MSVC linker, you might as well use the MSVC lib/dlltool.

dpaoliello commented:
Blocking to wait for <https://github.com/llvm/llvm-project/pull/84834>

cjacek commented:
I have similar feeling to @efriedma-quic . There is probably not much use of those fixes on their own. They are a step towards stand-alone toolchain and are required for the linker, both for its libs part and to generate test cases, but they alone are not enough. They will be more useful when paired with linker support, which I hope will be in a useful shape for LLVM 19. Until then MSVC is required anyway.

dpaoliello commented:
Sorry, I've got to disagree with this: this is blocking Rust's use of Arm64EC since it can generate import libraries on-the-fly using this functionality (rather than requiring that users install the Windows SDK).

I've tested Arm64EC with Rust, and this is the only blocking issue that I've found.

dpaoliello commented:
> So your use-case is basically equivalent to using llvm-dlltool, except not using the text parser?

Yep.

> If this is actually enough to make Rust targets usable, then I guess we could consider it, but the fixes aren't structured in a way to make it obvious this won't impact non-ARM64EC targets.

That's fair enough.

@cjacek you have more experience in this area than I do - what's your confidence level with backporting these changes?

cjacek commented:
It's great to hear that it's enough got Rust. I re-reviewed those commits for impact on non-ARM64EC target and I think it's safe enough.

One thing we could consider is to skip .def file parser part of it. The main downside of it would be that it would require skipping some tests too. Other than than that, it's unlikely to matter and I think that Rust calls `writeImportLibrary` directly anyway.

dpaoliello committed:
[Arm64EC] Copy import descriptors to the EC Map (#84834)

As noted in <https://github.com/llvm/llvm-project/pull/78537>, MSVC
places import descriptors in both the EC and regular map - that PR moved
the descriptors to ONLY the regular map, however this causes linking
errors when linking as Arm64EC:

```
bcryptprimitives.lib(bcryptprimitives.dll) : error LNK2001: unresolved external symbol __IMPORT_DESCRIPTOR_bcryptprimitives (EC Symbol)
```

This change copies import descriptors from the regular map to the EC
map, which fixes this linking error.

dpaoliello commented:
<https://github.com/llvm/llvm-project/pull/84834> has been merged into master and cherry-picked here, so this is ready to land.

> One thing we could consider is to skip .def file parser part of it. 

I'd rather not modify commits since they've already been tested in master and it makes subsequent cherry-picks more difficult.

dpaoliello commented:
@cjacek if you really want me to revert the def parser portion, then I can do that. Otherwise, can you please sign off?

cjacek reviewed (approved):
I think it's fine to keep it, thanks.

nikic commented on llvm/include/llvm/Object/COFFImportFile.h:
```diff
@@ -91,6 +120,10 @@ struct COFFShortExport {
 
+  /// Specifies EXPORTAS name. In a .def file, this is "bar" in
+  /// "EXPORTS\nfoo EXPORTAS bar".
+  std::string ExportAs;
```
This looks like an ABI break.

dpaoliello replied:
If that's problematic, this specific part of the change isn't needed for Rust's scenario.

tstellar replied:
If you can remove this part of the patch, that would be great.

dpaoliello replied:
Done

dpaoliello commented:
@tstellar I'd like to get this merged into the v18 release branch, but there's some discussion of whether we should trim the changes down to just what is needed or leave them as they were merged into master. Can you please let me know what your preference is?

dpaoliello committed:
Remove support for EXPORTAS in def files to maintain ABI compatibility for COFFShortExport
